Java: Add OpenFeign integration docs. (#3996)
getsentry/sentry-docs: src/platforms/java/performance/instrumentation/open-feign.mdx

---
title: OpenFeign Integration
sidebar_order: 20

> description: "Learn how to capture OpenFeign based HTTP clients performance"

Reviewer: ```suggestion
description: "Learn how to capture the performance of OpenFeign-based HTTP clients."
```

---

<Note>

Capturing transactions requires that you first <PlatformLink to="/performance/">set up performance monitoring</PlatformLink> if you haven't already.

</Note>

> Sentry OpenFeign integration provides `SentryFeignClient` that creates a span for each outgoing HTTP request executed with a [Feign](https://github.com/OpenFeign/feign) based HTTP client.

Reviewer: ```suggestion
Sentry OpenFeign integration provides the `SentryFeignClient`, which creates a span for each outgoing HTTP request executed with a [Feign](https://github.com/OpenFeign/feign)-based HTTP client.
```

### Install

```xml {tabTitle:Maven}
<dependency>
    <groupId>io.sentry</groupId>
    <artifactId>sentry-openfeign</artifactId>
    <version>{{ packages.version('sentry.java.openfeign', '5.1.0') }}</version>
</dependency>
```

```groovy {tabTitle:Gradle}
implementation 'io.sentry:sentry-openfeign:{{ packages.version('sentry.java.openfeign', '5.1.0') }}'
```

```scala {tabTitle: SBT}
libraryDependencies += "io.sentry" % "sentry-openfeign" % "{{ packages.version('sentry.java.openfeign', '5.1.0') }}"
```

For other dependency managers see the [central Maven repository](https://search.maven.org/artifact/io.sentry/sentry-openfeign).

## Configure

Add `SentryCapability` to Feign builder:

```java
import feign.Feign;
import io.sentry.openfeign.SentryCapability;

YourApi api = Feign.builder()
    .addCapability(new SentryCapability())
    ...
    .target(YourApi.class, "https://your-api-host/");
```

```kotlin
import feign.Feign
import io.sentry.openfeign.SentryCapability

val api = Feign.builder()
    .addCapability(SentryCapability())
    ...
    .target(YourApi::class.java, "https://your-api-host/")
```

> ## Modify or Drop spans

Reviewer: ```suggestion
## Modify or Drop Spans
```

```java
import feign.Feign;
import io.sentry.openfeign.SentryCapability;

YourApi api = Feign.builder()
    .addCapability(
        new SentryCapability(
            (span, request, response) -> {
              // modify span or return `null` to drop
              if (request.url().endsWith("/todos")) {
                span.setTag("tag-name", "tag-value");
              }
              return span;
            }))
    ...
    .target(YourApi.class, "https://your-api-host/");
```

```kotlin
import feign.Feign
import feign.Request
import feign.Response
import io.sentry.ISpan
import io.sentry.openfeign.SentryCapability
import io.sentry.openfeign.SentryFeignClient.BeforeSpanCallback

val api = Feign.builder()
    .addCapability(
        SentryCapability(
            BeforeSpanCallback { span: ISpan, request: Request, response: Response? ->
                // modify span or return `null` to drop
                if (request.url().endsWith("/todos")) {
                    span.setTag("tag-name", "tag-value")
                }
                span
            }))

>     .target(YourApi.class, "https://your-api-host/");

Reviewer 2: ```suggestion
    .target(YourApi::class.java, "https://your-api-host/")
```

for Kotlin, we need `::class.java` and no `;`

---
title: OpenFeign Integration
sidebar_order: 20
description: "Learn how to capture OpenFeign based HTTP clients performance"
---

<Note>

Capturing transactions requires that you first <PlatformLink to="/performance/">set up performance monitoring</PlatformLink> if you haven't already.

</Note>

Sentry OpenFeign integration provides `SentryFeignClient` that creates a span for each outgoing HTTP request executed with a [Feign](https://github.com/OpenFeign/feign) based HTTP client.

### Install

```xml {tabTitle:Maven}
<dependency>
    <groupId>io.sentry</groupId>
    <artifactId>sentry-openfeign</artifactId>
    <version>{{ packages.version('sentry.java.openfeign', '5.1.0') }}</version>
</dependency>
```

```groovy {tabTitle:Gradle}
implementation 'io.sentry:sentry-openfeign:{{ packages.version('sentry.java.openfeign', '5.1.0') }}'
```

```scala {tabTitle: SBT}
libraryDependencies += "io.sentry" % "sentry-openfeign" % "{{ packages.version('sentry.java.openfeign', '5.1.0') }}"
```

> For other dependency managers see the [central Maven repository](https://search.maven.org/artifact/io.sentry/sentry-openfeign).

Reviewer 2: ```suggestion
For other dependency managers see the [central Maven repository](https://search.maven.org/artifact/io.sentry/).
```

when I open https://search.maven.org/artifact/io.sentry/ it does not work though, the search link probably has changed and we'd have to fix across the docs

eg https://search.maven.org/search?q=g:io.sentry

Author: https://search.maven.org/artifact/io.sentry/sentry-openfeign is a correct link. As I understand, this links to a page where you can look up syntax for adding dependency to other dependency managers, so a link to exact package makes sense..? That's how it is for other modules.

Reviewer 2: for me it opens a blank page

Reviewer 2: ahhh because its not published yet, my bad hehe

import io.sentry.openfeign.SentryCapability;

YourApi api = Feign.builder()
    .addCapability(new SentryCapability())
    ...
    .target(YourApi.class, "https://your-api-host/");
```

```kotlin
import feign.Feign
import io.sentry.openfeign.SentryCapability

val api = Feign.builder()
    .addCapability(SentryCapability())
    ...
    .target(YourApi::class.java, "https://your-api-host/")
```

## Modify or Drop spans

Spans created around HTTP requests can be modified or dropped using `SentryFeignClient.BeforeSpanCallback` passed to `SentryCapability`:

```java
import feign.Feign;
import io.sentry.openfeign.SentryCapability;

YourApi api = Feign.builder()
    .addCapability(
        new SentryCapability(
            (span, request, response) -> {
              // modify span or return `null` to drop
              if (request.url().endsWith("/todos")) {
                span.setTag("tag-name", "tag-value");
              }
              return span;
            }))
    ...
    .target(YourApi.class, "https://your-api-host/");
```

```kotlin
import feign.Feign
import io.sentry.openfeign.SentryCapability

val api = Feign.builder()
    .addCapability(
        SentryCapability(

>             BeforeSpanCallback { span: ISpan, request: Request, response: Response? ->

Reviewer 2: does BeforeSpanCallback require importing as well?

Author: Plus few others 👍
